### ISSUE TITLE
Crash in Windows: UnitConvertor?

### ISSUE THREAD

**Issue by hekkup** [hrobeers/QtUnits#1]
I'm using QtUnits as static library (QTUNITS_STATIC defined) in Windows 10 with Qt 5.9.1, MS Visual C++ compiler 14.0 and Boost 1.66.0. QtUnits crashes. With debugging I figure there is some problem with internal usage of UnitConvertor in QuantityBase. I can see QuantityBase::convertorMap map has a correct number of convertors but as soon as UnitConvertor methods are called, a crash results.

Compiling unit tests in Windows doesn't work out of the box -- CMake has problems.

In Linux everything works right away.

Below is an example code in which the crash can be seen. Stand-alone UnitConvertor works and internal value can be accessed. Problem happens as soon as UnitConvertor is involved internally in QuantityBase.

```
MainWindow::MainWindow(QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    UnitConvertor<cgs::length, si::length> cmConv("cm");
    quantity<si::length, qreal> internalValue(0.5 * si::meter);
    qreal convertedCm = cmConv.fromInternalValue(internalValue);
    qDebug() << "Got converted value: " << convertedCm << cmConv.unitSymbol();

    quantity<si::length, qreal> length(0.35 * si::meter);
    Length displayLength(length, LengthUnit::cm);

    qDebug() << "length.value():" << length.value() << "m";
    qDebug() << "displayLength.internalValue().value():" << displayLength.internalValue().value();

    // crash:
    //qDebug() << "displayLength.value():" << QString::number(displayLength.value());

    // crash:
    //qDebug() << "displayLength.unitSymbol():" << displayLength.unitSymbol();
}
```

**Comment by hrobeers** [comment]
Hi hekkup, thanks for reporting this. I wasn't aware other people were using this lib.

I don't run windows myself and only compile it on windows sporadically.
If you find a solution to this problem, feel free to make a pull request.
Otherwise you'll have to wait until I get access to a windows machine an find some time to figure this out. It might take a few weeks until I get access to a windows 7 machine.

Also this lib is limited to length, area and angles so if you need more feel free to add them and make a PR.

**Comment by hekkup** [comment]
A workaround is to remove constexpr keyword from UnitConvertor constructor and its instantiations in Length.cpp, Area.cpp and Degree.cpp files. It looks like the convertors don't get instantiated properly with MSVC++ compiler. I got a clue about this when I put a debug message into the constructor and only got message on explicit instantiation. Yeah, MSVC++ compiler lets you do it while GCC won't.

Now to wait for the unintended consequences... I'm not familiar with constexpr keyword so I don't know how badly this breaks QtUnits.

I would rather not make a push request about this for the time being unless you give me green light; the reason is that I don't really know what I am doing.

**Comment by hrobeers** [comment]
This feels like a faulty implementation of constexpr in MSVC as I can't find a reason why this should not work.
Can you try if the following things work?

1. replace all "constexpr" keywords by "const" in length.cpp, area.cpp and angle.cpp ?
2. replace the "const" keyword in front of "insertconvertor" by "constexpr"

Please let me know if any of the above works.

It is safe to leave out the constexpr or const keywords but it's not as clean. So if none of the above works you can safely remove the keywords for your use. I'll update the code later.

Thanks for reporting.

**Comment by hrobeers** [comment]
I just tested on linux.
Replacing all "constexpr" by simple "const" in length.cpp, area.cpp and angle.cpp should work.
Can you try that?

**Comment by hekkup** [comment]
Tested and works.

> 1. replace all "constexpr" keywords by "const" in length.cpp, area.cpp and angle.cpp ?

This works.

You can leave constexpr in front of *_SYMBOL definitions.

You can also leave or change constexpr in front of UnitConvertor constructor -- I tested changing it to const, too, with no problem.

> 2. replace the "const" keyword in front of "insertconvertor" by "constexpr"

This won't compile when you do this either in insertConvertor() method in quantitybase.hpp or in quantity class files (length.cpp, area.cpp, angle.cpp).

**Comment by hrobeers** [comment]
Great, that was what I expected and hoped for.

Fixed by commit 8c927ad72e3f90901a0f362283ce5cc9c2b8c974
